File: problems/11/11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python3
#
# [11] Container With Most Water
#


# @lc code=start
class Solution:
    def maxArea(self, height: List[int]) -> int:
        # max amount of water container can store
        # height = [1,8,6,2,5,4,8,3,7]
        #             s             e
        # area -> min(height[l], height[r]) * (r - l)
        # 2 pointers
        # keep track of current max water
        # which ever side is shorter, move
        # time O(N), space O(1)

        n = len(height)
        l, r = 0, n - 1
        curr_max = 0
        while l < r:
            curr = min(height[l], height[r]) * (r - l)
            curr_max = max(curr, curr_max)

            if height[l] <= height[r]:
                l += 1
            else:
                r -= 1

        return curr_max


# @lc code=end

File: problems/11/11.container-with-most-water.java
/*
 * @lc app=leetcode id=11 lang=java
 *
 * [11] Container With Most Water
 */

// @lc code=start
class Solution {
    public int maxArea(int[] height) {
        /**
         * Find 2 lines that create the largest area rectangle
         *
         * intuition:
         * - 2 pointers
         * - greedily update l, r to find max area where
         * - if height[r] > height[l] -> increment l
         * - else -> decrement r
         * - continue till l == r, then return found max
         *
         * O(N) time, O(1) space
         */

        int l = 0;
        int r = height.length - 1;
        int currMax = 0;

        while (l < r) {
            int currArea = (r - l) * Math.min(height[l], height[r]);
            currMax = Math.max(currMax, currArea);

            if (height[r] > height[l]) {
                l++;
            } else {
                r--;
            }
        }

        return currMax;
    }
}
// @lc code=end
